### ISSUE TITLE
Open Reader View in an extension page

### ISSUE THREAD

**Issue by csadilek** [mozilla-mobile/android-components#2985]
When activating reader view, we are currently transforming the existing document in our web extension content script. 

This leaves two problems:

- We don't get a history entry for reader view (although we could use pushState and other approaches see https://github.com/mozilla-mobile/android-components/issues/2879 for details)

- We effectively can't isolate reader view from JavaScript that was loaded with the original page e.g. a setTimeout call that shows a popup / ad etc.

A solution to this would be to open reader view in an [extension page](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/user_interface/Extension_pages). So, our content script would send a message to a background script, which would call `browser.tabs.update()`. We can then transform the DOM using a script loaded by the extension page. We can pass the original document along in the message, and follow Fennec's approach and download the document again (using XHR) as a fallback (e.g. when the extension page is navigated to directly, see also https://github.com/mozilla-mobile/android-components/issues/2980).

Another big advantage of this approach is that we would no longer need to inject all scripts e.g.`Readability.js` into every single page. We only need to inject the functionality that checks whether or not the page is readerable. The remaining scripts only need to be loaded by the extension page once Reader View is active.

In order to make this work, we need support for `browser.tabs.update()` for pages packaged in the web extension. This currently doesn't seem to work in GV: https://bugzilla.mozilla.org/show_bug.cgi?id=1550877

**Status - Remaining GV issues we're blocked on / still need fixed and uplifted to beta:**

* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1554302 (no history entry for extension pages)~~
* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1554302 (`browser.tabs.update` only works once per session)~~
* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1554260 (WebExtension Page messages are sent to the Runtime instead of GeckoSessions)~~
* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1553371 (WebExtension pages don't have access to privileged APIs)~~
* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1551377 (WebExtension `tabId` should return a unique ID for each tab)~~

Other related GV issues, but not blockers:
* [x] https://bugzilla.mozilla.org/show_bug.cgi?id=1634423 (capturePixels returns empty bitmap for extension pages)
* [x] https://bugzilla.mozilla.org/show_bug.cgi?id=1634428 (Intermediate loads of about:blank when navigating to/from extension page)
* [ ] https://bugzilla.mozilla.org/show_bug.cgi?id=1550144 (Provide API to add custom history entry (e.g. about:reader?url=))
* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1560048 (Expose internal web extension ID)~~
* [x] ~~https://bugzilla.mozilla.org/show_bug.cgi?id=1551278 (Content messaging delegate receives messages from all extensions)~~

**Comment by csadilek** [comment]
Filed https://bugzilla.mozilla.org/show_bug.cgi?id=1550877 for the `browser.tabs.update` issue.

**Comment by pwd-github** [comment]
One of the pain points of Reader View is that it's missing from history (on both desktop and Fennec) and thus it's really hard to follow along when you use it for reading something that's paginated, but I believe it would be far more jarring to have a bunch of additional tabs open separate from the page content. As a user, I certainly don't want Reader View to create tab management issues.

**Comment by csadilek** [comment]
@pwd-github agreed. the change happening here is **not** about opening an additional tabs. It's about using `browser.tabs.update` to load reader view as a new page in the current tab with the benefits described above.

**Comment by csadilek** [comment]
Updated description with status of pending GV bugs which we need fixed and uplifted to beta.

**Comment by bifleming** [comment]
@vesta0 Post MVP?

**Comment by vesta0** [comment]
@bifleming this isn't marked as an MVP blocker. It's P1 as soon as it's unblocked.

**Comment by csadilek** [comment]
This is no longer blocked now and could be worked on (All GV functionality landed).

**Comment by csadilek** [comment]
Found two more problems with extension pages in GV:
* https://bugzilla.mozilla.org/show_bug.cgi?id=1634423 (capturePixels returns empty bitmap for extension pages)
* https://bugzilla.mozilla.org/show_bug.cgi?id=1634428 (Intermediate loads of about:blank when navigating to/from extension page)

I don't think we need to block on them though before landing a fix here.
